### PR TITLE
Admin Portal

### PR DESCRIPTION
**Pull request by jim-minter**
Ready for review.

### CODE REVIEW

**Comment by jim-minter** [comment]
@mjudeikis could you take a look through this please?

File: cmd/aro/portal.go [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	if err != nil {
+		return err
+	}
+
+	cipher, err := encryption.NewXChaCha20Poly1305(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	dbc, err := database.NewDatabaseClient(ctx, log.WithField("component", "database"), _env, &noop.Noop{}, cipher)
```
Comment (mjudeikis) [inline cmd/aro/portal.go:88]: Any reason we are not emitting metrics for portal db queries? This creates dark spots in our metrics and database view.

File: hack/devtools/deploy-shared-env.sh [hunk 2]
```diff
@@ -86,6 +86,14 @@ import_certs_secrets() {
         --vault-name "$KEYVAULT_PREFIX-svc" \
         --name rp-server \
         --file secrets/localhost.pem
+    az keyvault certificate import \
+        --vault-name "$KEYVAULT_PREFIX-por" \
+        --name portal-server \
+        --file secrets/localhost.pem
```
Comment (mjudeikis) [inline hack/devtools/deploy-shared-env.sh:92]: This was intentionally reused in dev?

File: pkg/deploy/deploy.go [hunk 3]
```diff
@@ -179,6 +179,11 @@ func (d *deployer) configureDNS(ctx context.Context) error {
 		return err
 	}
 
+	portalPip, err := d.publicipaddresses.Get(ctx, d.config.ResourceGroupName, "portal-pip", "")
```
Comment (mjudeikis) [inline pkg/deploy/deploy.go:182]: Did you tested any of this deploy code or this is "fix in prod" case?

File: pkg/portal/kubeconfig/kubeconfig.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	if err != nil {
+		k.internalServerError(w, err)
+		return
+	}
+
+	b, err := k.makeKubeconfig("https://"+r.Host+resourceID+"/kubeconfig/proxy", token)
+	if err != nil {
+		k.internalServerError(w, err)
+		return
+	}
```
Comment (mjudeikis) [inline pkg/portal/kubeconfig/kubeconfig.go:130]: If this fails, does not means we gonna be leaking portal documents?

File: pkg/portal/util/clientcache/clientcache.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+func (c *clientCache) expire() {
+	now := c.now()
+	for k, v := range c.m {
+		if now.After(v.expires) {
+			v.cli.CloseIdleConnections()
+			delete(c.m, k)
+		}
+	}
+}
+
```
Comment (mjudeikis) [inline pkg/portal/util/clientcache/clientcache.go:51]: Does this implies that if we dont call any Get, Put, clients will not be expired? Potentially sitting inmemory?

File: pkg/portal/prometheus/proxy.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+			if attr.Namespace == "" && attr.Key == "src" && strings.HasPrefix(n.Attr[i].Val, "/") {
+				n.Attr[i].Val = "." + n.Attr[i].Val
+			}
+		}
+
+		if len(n.Attr) == 0 {
+			n.FirstChild.Data = strings.Replace(n.FirstChild.Data, `var PATH_PREFIX = "";`, `var PATH_PREFIX = ".";`, 1)
+		}
+	}
+}
```
Comment (mjudeikis) [inline pkg/portal/prometheus/proxy.go:152]: All this is asking for explanation comments

File: pkg/portal/prometheus/proxy.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+func walk(n *html.Node, f func(*html.Node)) {
+	f(n)
+
+	for c := n.FirstChild; c != nil; c = c.NextSibling {
+		walk(c, f)
```
Comment (mjudeikis) [inline pkg/portal/prometheus/proxy.go:158]: Same as above.

File: pkg/portal/ssh/proxy.go [hunk 8]
```diff
@@ -0,0 +1,334 @@
+package ssh
```
Comment (mjudeikis) [inline pkg/portal/ssh/proxy.go:1]: All this file is :exploding_head: . Would appreciate some comments and a walkthrough again :/

**Review by mjudeikis — COMMENTED** [review]
apart of networking shenanigans looks goodish...

File: hack/gendiscoverycache/genrbac.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+		switch gr {
+		case "tokenreviews.authentication.k8s.io",
+			"selfsubjectaccessreviews.authorization.k8s.io",
+			"selfsubjectrulesreviews.authorization.k8s.io",
+			"subjectaccessreviews.authorization.k8s.io",
+			"resourceaccessreviews.authorization.openshift.io",
+			"subjectaccessreviews.authorization.openshift.io":
+			return true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply hack/gendiscoverycache/genrbac.go:133]: These objects are k8s' equivalent of the proc filesystem - they're not actually persisted to etcd.  Create is kind-of like read here.  I'm broadly aiming to match what cluster-reader does.

File: cmd/aro/portal.go (hunk 1 above)
Comment (jim-minter) [reply cmd/aro/portal.go:88]: Good catch, will fix.

File: hack/devtools/deploy-shared-env.sh (hunk 2 above)
Comment (jim-minter) [reply hack/devtools/deploy-shared-env.sh:92]: Yes.

File: pkg/deploy/deploy.go (hunk 3 above)
Comment (jim-minter) [reply pkg/deploy/deploy.go:182]: No - want to try INT this week.

File: pkg/portal/kubeconfig/kubeconfig.go (hunk 4 above)
Comment (jim-minter) [reply pkg/portal/kubeconfig/kubeconfig.go:130]: No: there is a TTL set on the Portal collection so they will auto-expire.

File: pkg/portal/util/clientcache/clientcache.go (hunk 5 above)
Comment (jim-minter) [reply pkg/portal/util/clientcache/clientcache.go:51]: Yes.  Do you think that's adequate?

File: hack/gendiscoverycache/genrbac.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (mjudeikis) [reply hack/gendiscoverycache/genrbac.go:133]: TIL

File: pkg/portal/kubeconfig/kubeconfig.go (hunk 4 above)
Comment (mjudeikis) [reply pkg/portal/kubeconfig/kubeconfig.go:130]: ack

File: pkg/portal/util/clientcache/clientcache.go (hunk 5 above)
Comment (mjudeikis) [reply pkg/portal/util/clientcache/clientcache.go:51]: It does the job. I just never seen an implementation like this. Usually, its some goroutine running and checking expiring things. Basically, I'm worried if somebody can access the memory of the VM (one more side-channel or smth) we could have stale clients in regions where we run 1-2 clusters for months... But at the same time if somebody dumped the memory - we have bigger issues :man_shrugging:  I think I'm fine with this as long as we are aware

**Comment by jim-minter** [comment]
@mjudeikis review feedback addressed with additional commits, ptal again.

**Comment by jim-minter** [comment]
Tested in INT - merging!

**Pull request merged**